--- a/llm/generate/gen_linux.sh
+++ b/llm/generate/gen_linux.sh
@@ -52,8 +52,7 @@
 COMMON_CMAKE_DEFS="-DCMAKE_POSITION_INDEPENDENT_CODE=on -DLLAMA_NATIVE=off -DLLAMA_AVX=on -DLLAMA_AVX2=off -DLLAMA_AVX512=off -DLLAMA_FMA=off -DLLAMA_F16C=off"
 source $(dirname $0)/gen_common.sh
 init_vars
-git_module_setup
-apply_patches
+patch ${LLAMACPP_DIR}/examples/server/CMakeLists.txt < @cmakelistsPatch@
 
 if [ -z "${OLLAMA_SKIP_CPU_GENERATE}" ]; then
     # Users building from source can tune the exact flags we pass to cmake for configuring
@@ -186,5 +185,3 @@
     #       them being present at runtime on the host
     compress_libs
 fi
-
-cleanup
